examples/thermostat/thermostat-common/src/atomic-write-manager.cpp

```diff
@@ -0,0 +1,535 @@
+void ThermostatAtomicWriteManager::ScheduleTimer(EndpointId endpoint, System::Clock::Milliseconds16 timeout)
+{
+    DeviceLayer::SystemLayer().StartTimer(timeout, TimerExpiredCallback,
+                                          reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+/**
+ * @brief Clears the currently scheduled timer.
+ *
+ * @param[in] endpoint The endpoint to use.
+ */
+void ThermostatAtomicWriteManager::ClearTimer(EndpointId endpoint)
+{
+    DeviceLayer::SystemLayer().CancelTimer(TimerExpiredCallback, reinterpret_cast<void *>(static_cast<uintptr_t>(endpoint)));
+}
+
+void ThermostatAtomicWriteManager::OnTimerExpired(EndpointId endpoint)
+{
+    for (size_t i = 0; i < ArraySize(mAtomicWriteSessions); ++i)
+    {
+
+        auto atomicWriteState = mAtomicWriteSessions[i];
+        if (atomicWriteState.state == kAtomicWriteState_Open && atomicWriteState.endpointId == endpoint)
+        {
+            if (mDelegate != nullptr)
+            {
+                mDelegate->OnRollbackWrite(endpoint, Presets::Id);
+                mDelegate->OnRollbackWrite(endpoint, Schedules::Id);
+            }
+            ResetWrite(endpoint);
+        }
+    }
+}
+
+void ThermostatAtomicWriteManager::SetWriteState(EndpointId endpoint, ScopedNodeId originatorNodeId, AtomicWriteState state)
+{
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        mAtomicWriteSessions[ep].state      = state;
+        mAtomicWriteSessions[ep].endpointId = endpoint;
+        mAtomicWriteSessions[ep].nodeId     = originatorNodeId;
+    }
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId, EndpointId endpoint)
+{
+    // These are the only two atomic attributes
+    if (attributeId.has_value() && attributeId.value() != Presets::Id && attributeId.value() != Schedules::Id)
+    {
+        return false;
+    }
+
+    bool inAtomicWrite = false;
+    uint16_t ep =
+        emberAfGetClusterServerEndpointIndex(endpoint, Thermostat::Id, MATTER_DM_THERMOSTAT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (ep < ArraySize(mAtomicWriteSessions))
+    {
+        inAtomicWrite = mAtomicWriteSessions[ep].state == kAtomicWriteState_Open;
+    }
+    return inAtomicWrite;
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId,
+                                           const Access::SubjectDescriptor & subjectDescriptor, EndpointId endpoint)
+{
+    if (!InWrite(attributeId, endpoint))
+    {
+        return false;
+    }
+    return subjectDescriptor.authMode == Access::AuthMode::kCase &&
+        GetAtomicWriteScopedNodeId(attributeId, endpoint) == ScopedNodeId(subjectDescriptor.subject, subjectDescriptor.fabricIndex);
+}
+
+bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId, CommandHandler * commandObj,
```

Comment:
like i commented before looks like both apis with SubjectDescriptor and commandobj check the scoped node id so maybe remove both and add an api with scopednodeid as the argument

Reply:
As I replied before, the subject descriptor is acquired and validated differently for commands and attribute accesses; abstracting that would just move this logic elsewhere and require exporting a method with CommandHandler anyway. I'm going to stick with having this logic in one place instead of three.

Reply:
why would you need to export a method for CommandHandler. i do not understand this part. you have 2 utility functions that get the scoped node id from either the commandhandler object or subjectdescriptor object in the callee for these methods. and your api inwrite would look like - bool ThermostatAtomicWriteManager::InWrite(const std::optional<AttributeId> attributeId, ScopedNodeId nodeId, EndpointId endpoint) instead of 2 versions with commandobject and subjectdescriptor. Am i missing something here?

Reply:
Because this would mean replacing two methods with three methods:
```
bool InWrite(const std::optional<AttributeId> attributeId, const ScopedNodeId & nodeId,
                         EndpointId endpoint);
ScopedNodeId GetScopedNodeIdFromSubjectDescriptor(const Access::SubjectDescriptor & subjectDescriptor);                 
ScopedNodeId GetScopedNodeIdFromCommandHandler(const CommandHandler * commandObj);   
```
and now requiring callers to call two methods instead of one:
```
awm->InWrite(aPath.mAttributeId, awm->GetScopedNodeIdFromSubjectDescriptor(subjectDescriptor), aPath.mEndpointId)
awm->InWrite(aPath.mAttributeId, awm->GetScopedNodeIdFromCommandHandler(commandObj), aPath.mEndpointId)
```
as well as opening up the ability to work around the security check by crafting their own ScopedNodeId.

Reply:
I think the key thing is this: finding the right ScopedNodeId is not trivial, and we should not be pushing that work off on applications if we can help it.  But I'm waiting to see what things look like with more of the logic just moved into the cluster.